### ISSUE TITLE
[Question] Volume faulted after reboot, need manually salvage?

### ISSUE THREAD

**Issue by LuizBuffa25** [longhorn/longhorn#2309]
**Question**
Every time a reboot my node (using only 1 replica), my volume became "faulted", every time I need to go to UI and perform a salvage, my question is: There is a way to do it automatically or via command line?

* already set auto-salvage: true on default-setting.yaml, but it does not work.

**Environment:**
Longhorn version: 1.1.0
Kubernetes version: 1.20

 - Node config
   - OS type and version
   - CPU per node:
   - Memory per node:
   - Disk type
   - Network bandwidth and latency between the nodes:
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal):

**Additional context**

**Comment by shuo-wu** [comment]
The values in `default-setting.yaml` will be applied to newly launched Longhorn systems only. In other words, if you re-deploy/update this YAML files for an existing Longhorn system, it doesn't work.

Here, you need to directly check the setting value via Longhorn UI or `kubectl -n longhorn-system get settings`

**Comment by LuizBuffa25** [comment]
Also, I reboot my server and $ kubectl get pods -n longhorn-system, after reboot longhorn try to create "shared-manager-MY-VOL-NAME" but the pod went to terminating and don't restart anymore.

**Comment by joshimoo** [comment]
@LuizBuffa25 you linked to the v1.0 deployment, can you check if you actually used the correct v1.1.0 deployment files?
https://raw.githubusercontent.com/longhorn/longhorn/v1.0.0/deploy/longhorn.yaml

**Comment by LuizBuffa25** [comment]
@joshimoo Just a copy and paste mistake, I am using v1.1.0.

**Comment by shuo-wu** [comment]
OK. Can you check if the result of `kubectl -n longhorn-system get settings auto-salvage` is true? If NO, then something may be wrong. And it's better to generate a support bundle then we can take a deep look.

I guess you are using a RWX Longhorn volume. Since the volume is faulted, the share manager pod should be stopped. And it will be created/recreated after the volume attachment/reattachment. Typically you don't need to take care of that pod.

**Comment by LuizBuffa25** [comment]
output:
NAME           VALUE   AGE
auto-salvage   true    24h

Yes, I am using RWX. There is a workaround to "salvage" using the command line instead of UI?

**Comment by shuo-wu** [comment]
OK, then there may be a bug for this feature. Can you provide the support bundle? So that there is a chance to fix it and you don't need to worry about the manual salvage.

Using the command line to salvage replicas manually is possible but a little bit complicated. You need to figure out all failed replicas in running disks, then pick up the possible valid replicas and unset field `spec.failedAt`.
```
kubectl -n longhorn-system get replicas --output=jsonpath="{.items[?(@.status.volumeName==\"<THE VOLUME NAME YOU ARE CHECKING>\")].metadata.name}"
kubectl -n longhorn-system edit replicas
```

**Comment by LuizBuffa25** [comment]
Hey just one more question, when I submit a support bundle, any data of my longhorn will be exposed?

**Comment by shuo-wu** [comment]
It seems that you just upgraded Longhorn before you generate the support bundle. I didn't saw any salvage related logs in it.

**Comment by LuizBuffa25** [comment]
Yes, I will simulate the error later and send another support bundle.

**Comment by LuizBuffa25** [comment]
Sent an e-mail with a new support bundle.

**Comment by ch8zer** [comment]
I'm experiencing the exact same thing. I'm running on k3s - fresh install. I've emailed in a support bundle as well - hope it helps.

**Comment by Mist-Hunter** [comment]
Same issue. Ubuntu 20.04.2 LTS, Longhorn 1.10.1, Rancher 2.5.7, K3s 1.20, RWX volume.

**Comment by guangbochen** [comment]
similar issue here, when the host is rebooted the minio pod recovery is blocked by the salvage. cc @yasker 
```
7m34s       Warning   FailedAttachVolume       pod/minio-6ccfd78c56-q8h9z    AttachVolume.Attach failed for volume "pvc-41383415-f0ab-4b92-82d7-a136c01d903e" : rpc error: code = Aborted desc = The volume pvc-41383415-f0ab-4b92-82d7-a136c01d903e in state detached is not ready for workloads, is waiting on controller false
```
error log from the longhorn-manager:
```
time="2021-03-26T06:44:58Z" level=info msg="All replicas are failed, set engine salvageRequested to true" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=gc-harv2 owner=gc-harv2 state=detached volume=pvc-41383415-f0ab-4b92-82d7-a136c01d903e
time="2021-03-26T06:44:58Z" level=warning msg="Error syncing Longhorn volume longhorn-system/pvc-41383415-f0ab-4b92-82d7-a136c01d903e" controller=longhorn-volume error="fail to sync longhorn-system/pvc-41383415-f0ab-4b92-82d7-a136c01d903e: fail to reconcile volume state for pvc-41383415-f0ab-4b92-82d7-a136c01d903e: no node name provided to check node down or deleted" node=gc-harv2
time="2021-03-26T06:44:58Z" level=info msg="All replicas are failed, set engine salvageRequested to true" accessMode=rwo controller=longhorn-volume frontend=blockdev migratable=false node=gc-harv2 owner=gc-harv2 state=detached volume=pvc-41383415-f0ab-4b92-82d7-a136c01d903e
time="2021-03-26T06:44:58Z" level=warning msg="Error syncing Longhorn volume longhorn-system/pvc-41383415-f0ab-4b92-82d7-a136c01d903e" controller=longhorn-volume error="fail to sync longhorn-system/pvc-41383415-f0ab-4b92-82d7-a136c01d903e: fail to reconcile volume state for pvc-41383415-f0ab-4b92-82d7-a136c01d903e: no node name provided to check node down or deleted" node=gc-harv2
```
and the volume spec:
```
➜  ~ ks get volume pvc-41383415-f0ab-4b92-82d7-a136c01d903e -oyaml
apiVersion: longhorn.io/v1beta1
kind: Volume
metadata:
  creationTimestamp: "2021-03-25T06:40:48Z"
  finalizers:
  - longhorn.io
  generation: 3
  labels:
    longhornvolume: pvc-41383415-f0ab-4b92-82d7-a136c01d903e
  name: pvc-41383415-f0ab-4b92-82d7-a136c01d903e
  namespace: longhorn-system
  resourceVersion: "634269"
  uid: 77a7426a-eaa8-4462-bc7c-fa074c1fc038
spec:
  Standby: false
  accessMode: rwo
  backingImage: ""
  baseImage: ""
  dataLocality: disabled
  disableFrontend: false
  diskSelector: null
  engineImage: longhornio/longhorn-engine:v1.1.1-preview1
  fromBackup: ""
  frontend: blockdev
  lastAttachedBy: ""
  migratable: false
  migrationNodeID: ""
  nodeID: ""
  nodeSelector: null
  numberOfReplicas: 3
  recurringJobs: null
  revisionCounterDisabled: false
  size: "85899345920"
  staleReplicaTimeout: 30
```

**Comment by joshimoo** [comment]
@guangbochen I am working on a fix for your issue as part of a fix for this:
https://github.com/longhorn/longhorn/issues/2329#issuecomment-807527079
This is caused by the volume never being able to to complete a sync loop once the nodeID's have been unset.

**Comment by joshimoo** [comment]
Reproduce steps:
- create volume + pod with single replica + data locality
- turn off node

Problem 01: kubernetes + longhorn out of sync
- before csi changes, kubernetes and longhorn would go out of sync and kubernetes could no longer attach the volume since it's still desired to be attached to the prior node v.spec.NodeID is not unset.

Problem 02: salvage deadloop preventing volume updates
- after csi changes or let's say you manually unset the v.spec.NodeID via a detach from the UI or manual salvage of the replica while it's still on the downed node (manual salvage will first unset v.spec.nodeID) and then process the replicas (this operation will fail since the replica is down)
- the volume will now constantly fail to update since the isNodeDownCheck(v.spec.NodeID) https://github.com/longhorn/longhorn-manager/blob/master/controller/volume_controller.go#L1033 will always return error


Test Setup:
This fixes issues with salvage of multiple replicas as well, it's just easier to test with a single replica case.
- This should be tested with data locality enabled/disabled.
- This should be tested with replicaset / statefulset
- This should be tested with RWO/RWX volumes

Test steps:
- create RWO|RWX volume with replica count = 1 & data locality = enabled|disabled
- create deployment|statefulset for volume
- power down node of volume/replica
- wait for pod deletion & recreation
- volume will fail to attach since volume is not ready (i.e remains faulted, since single replica is on downed node)
- power up node
- verify auto salvage finishes (i.e pod completes start)
- verify volume attached & accessible by pod (i.e test data is available)

**Comment by taxilian** [comment]
I'm having this issue as well, though in my case the nodes didn't go down just the control plane endpoint load balancer did. I see code has been merged; is there a way to get it? I could provide a support bundle if that'd be useful

**Comment by shuo-wu** [comment]
> I'm having this issue as well, though in my case the nodes didn't go down just the control plane endpoint load balancer did. I see code has been merged; is there a way to get it? I could provide a support bundle if that'd be useful

The fix will be in the next release. For now, you may need to do manual salvage via Longhorn UI as a workaround. The manual salvage does what you mentioned in the Slack channel (unset `replica.spec.failedAt`)

**Comment by taxilian** [comment]
Interesting and thanks for the update; as an FYI, the manual salvage via longhorn UI did not do anything, but editing the replicas manually did.

**Comment by shuo-wu** [comment]
> Interesting and thanks for the update; as an FYI, the manual salvage via longhorn UI did not do anything, but editing the replicas manually did.

Does your Longhorn UI work fine when you issue other operations? Some users mentioned that the UI doesn't work with specific configs.

**Comment by taxilian** [comment]
Most other things are working in my UI, though backups sometimes don't show up

**Comment by shuo-wu** [comment]
Backup not shown may be caused by this issue: https://github.com/longhorn/longhorn/issues/1955. It will be fixed in v1.1.1 as well. But the salvage should work in UI... Can you provide more info if it still doesn't work after refreshing your UI page?

**Comment by khushboo-rancher** [comment]
Verified with Longhorn Master `04/07/2021`

Validation - **Pass**

The test cases from https://github.com/longhorn/longhorn/issues/2329#issuecomment-812134862 work fine. Validated this with the setting `Pod Deletion Policy When Node is Down` also, the pods with RWO were able to move to another node and got attached to volume but RWX remains on the same node with unavailable status.

**Comment by khushboo-rancher** [comment]
Closing this now based on the validation. @taxilian The fix will be available with Longhorn v1.1.1, feel free to reopen if you see the issue in future.
